### ISSUE TITLE
bf16 & tf32，can they be used together 

### ISSUE THREAD

**Issue by DSXiangLi** [tatsu-lab/stanford_alpaca#252]
In the hyperparameter, I saw the bf16 and tf32 are both set to true. However I don't they can be used together?

**Comment by hujunchao** [comment]
I have the same question

**Comment by dotsnangles** [comment]
I was searching for the internet to find the answer. and then I found this thread.
is it possible though? I think the latter arg probably overwrites the former.
if someone could clarity this, I would be grateful.

**Comment by yxchng** [comment]
any new update on this?

**Comment by tdolega** [comment]
By setting bf16, the model weights are saved in bf16, but gradients (computed in half-precision but converted to full-precision for the update) and optimizations (optimizer states) are still done in full 32-bit precision. By additionally setting tf32, you speed up that very calculations at the expense of precision, but since ultimately those weights are saved as bf16 anyway, it _probably_ has no effect on accuracy.

**Comment by mrs83** [comment]
I've been experimenting with bf16 and tf32 used together by fine-tuning a small model (Qwen2.5-Coder-0.5B) in a simulated federated learning setup. It's been super helpful for trying out different hyperparameters without burning through tons of resources: [FlowerTune-Qwen2.5-Coder-0.5B-Instruct](https://github.com/ethicalabs-ai/FlowerTune-Qwen2.5-Coder-0.5B-Instruct)
